src/CabanaPD_ForceModelsMulti.hpp

```diff
@@ -12,46 +12,211 @@
+struct FirstModelWithFractureTypeImpl<FractureType, ParameterPackType, Index,
+                                      false>
+{
+    using type = typename FirstModelWithFractureTypeImpl<
+        FractureType, ParameterPackType, Index + 1,
+        std::is_same_v<typename ParameterPackType::template value_type<
+                           Index + 1>::fracture_type,
+                       FractureType>>::type;
+};
+
+template <typename FractureType, typename ParameterPackType>
+struct FirstModelWithFractureType
+{
+    using type = typename FirstModelWithFractureTypeImpl<
+        FractureType, ParameterPackType, 0,
+        std::is_same_v<
+            typename ParameterPackType::template value_type<0>::fracture_type,
+            FractureType>>::type;
 };
 
 // Wrap multiple models in a single object.
-// TODO: this currently only supports bi-material systems.
-template <typename MaterialType, typename ModelType1, typename ModelType2,
-          typename ModelType12>
-struct ForceModels
+template <typename MaterialType, typename Indexing, typename ParameterPackType,
+          typename OutsideRangeFunctorType, typename Sequence>
+struct ForceModelsImpl
+{
+    static_assert( !std::is_same_v<Sequence, Sequence>,
+                   "ForceModelsImpl: We should never end up here" );
+};
+
+template <typename MaterialType, typename Indexing, typename ParameterPackType,
+          typename OutsideRangeFunctorType, std::size_t... Indices>
+struct ForceModelsImpl<MaterialType, Indexing, ParameterPackType,
+                       OutsideRangeFunctorType, std::index_sequence<Indices...>>
 {
     using material_type = MultiMaterial;
 
-    using first_model = ModelType1;
-    using model_type = typename first_model::model_type;
-    using base_model = typename first_model::base_model;
-    using thermal_type = typename first_model::thermal_type;
-    using fracture_type = typename first_model::fracture_type;
-
-    static_assert( ( std::is_same<typename ModelType1::thermal_type,
-                                  TemperatureIndependent>::value ||
-                     std::is_same<typename ModelType2::thermal_type,
-                                  TemperatureIndependent>::value ||
-                     std::is_same<typename ModelType12::thermal_type,
-                                  TemperatureIndependent>::value ),
-                   "Thermomechanics does not yet support multiple materials!" );
-
-    ForceModels( MaterialType t, const ModelType1 m1, ModelType2 m2,
-                 ModelType12 m12 )
+    static_assert(
+        (std::conjunction_v<
+            std::is_same<typename ParameterPackType::template value_type<
+                             Indices>::base_model,
+                         typename ParameterPackType::template value_type<
+                             0>::base_model>...>),
+        "All models need the same base model" );
+    // TODO this is a limitation of the solver we should get rid of
```

Comment:
I think it's reasonable to require matching base models for now; the issue is that both the `Force` iteration is different for base models and the `ForceModel` themselves do not necessarily have all the needed methods to work across base models